File: packages/mccomponents/libmccomponents/kernels/sample/ConstantvQEKernel.cc
// -*- C++ -*-
//
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//
//                                   Jiao Lin
//                      California Institute of Technology
//                      (C) 2006-2014  All Rights Reserved
//
// {LicenseText}
//
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//


#include <cmath>
#include "mccomponents/kernels/sample/ConstantvQEKernel.h"
#include "mccomponents/exception.h"
#include "mcni/math/number.h"
#include "mcni/neutron/units_conversion.h"
#include "mccomponents/math/random.h"


#ifdef DEBUG
#include "journal/debug.h"
#endif


struct mccomponents::kernels::ConstantvQEKernel::Details {

#ifdef DEBUG
  const static char jrnltag[];
  journal::debug_t debug;
  Details() : debug( jrnltag ) {}
#endif
};


#ifdef DEBUG
const char mccomponents::kernels::ConstantvQEKernel::Details::jrnltag[] = "ConstantvQEKernel";
#endif


mccomponents::kernels::ConstantvQEKernel::ConstantvQEKernel
( double Qx, double Qy, double Qz, double E, double dE,
  double absorption_coefficient,
  double scattering_coefficient)
  : m_Q(Qx, Qy, Qz), m_E(E), m_dE(dE),
    m_absorption_coefficient( absorption_coefficient ),
    m_scattering_coefficient( scattering_coefficient ),
    m_details( new Details )
{}


double
mccomponents::kernels::ConstantvQEKernel::absorption_coefficient(const mcni::Neutron::Event & ev )
{
  float_t v = ev.state.velocity.length();
  return m_absorption_coefficient * (2200/v);
}


double
mccomponents::kernels::ConstantvQEKernel::scattering_coefficient(const mcni::Neutron::Event & ev )
{
  return m_scattering_coefficient;
}


void
mccomponents::kernels::ConstantvQEKernel::absorb
( mcni::Neutron::Event & ev )
{
}


void
mccomponents::kernels::ConstantvQEKernel::S
( mcni::Neutron::Event & ev )
{
  namespace conversion = mcni::neutron_units_conversion;

  // input neutron state
  mcni::Neutron::State & state = ev.state;
  // incident neutron velocity
  double vi = state.velocity.length();
  // incident neutron energy
  double Ei = conversion::v2E( vi );
  // incident neutron mommentum
  Q_t ki = state.velocity * conversion::v2k;
  // output neutron momentum
  Q_t kf = ki - m_Q;
  // output neutron velocity
  state.velocity = kf * conversion::k2v;
  // final energy
  double Ef = conversion::v2E(state.velocity.length());
  // energy transfer
  double E = Ei - Ef;
  // 
  double dE = E-m_E, t = dE/m_dE;
  
  // adjust probability of neutron event
  ev.probability *= std::exp(-t*t/2);
}


// version
// $Id$

// End of file 

File: packages/mccomponents/python/mccomponents/sample/kernelxml/parser/ConstantvQEKernel.py
#!/usr/bin/env python
#
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#
#                                 Jiao Lin   
#                      California Institute of Technology
#                      (C) 2006-2014  All Rights Reserved
#
# <LicenseText>
#
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#


from .KernelNode import KernelNode as base


class ConstantvQEKernel(base):


    tag = "ConstantvQEKernel"

    def createKernel( self, **kwds ):
        from mccomponents.sample import constantvQEKernel
        E = self._parse( kwds['energy-transfer'] )
        dE = self._parse( kwds['dE'] )
        Q = self._parse( kwds['momentum-transfer'] )
        
        absorption_coefficient = kwds.get('absorption_coefficient')
        if absorption_coefficient:
            absorption_coefficient = self._parse(absorption_coefficient)
            
        scattering_coefficient = kwds.get('scattering_coefficient')
        if scattering_coefficient:
            scattering_coefficient = self._parse(scattering_coefficient)
            
        return constantvQEKernel(
            Q, E, dE, absorption_coefficient, scattering_coefficient)
    
    
    pass # end of ConstantvQEKernel


# version
__id__ = "$Id: ConstantvQEKernel.py 1797 2014-03-21 20:02:20Z linjiao $"

# End of file 
